integration-test/common/src/test/java/com/tencent/rss/test/QuorumTest.java

```diff
@@ -0,0 +1,194 @@
+
+  private static final String EXPECTED_EXCEPTION_MESSAGE = "Exception should be thrown";
+  private static ShuffleServerInfo shuffleServerInfo1;
+  private static ShuffleServerInfo shuffleServerInfo2;
+  private static ShuffleServerInfo shuffleServerInfo3;
+  private ShuffleWriteClientImpl shuffleWriteClientImpl;
+  private ShuffleWriteClientImpl shuffleWriteClientImpl2;
+
+  @BeforeClass
+  public static void setupServers() throws Exception {
+    CoordinatorConf coordinatorConf = getCoordinatorConf();
+    createCoordinatorServer(coordinatorConf);
+
+    ShuffleServerConf shuffleServerConf = getShuffleServerConf();
+    shuffleServerConf.setLong("rss.server.app.expired.withoutHeartbeat", 4000);
+    shuffleServerConf.setLong("rss.server.app.expired.withoutHeartbeat", 4000);
+    File tmpDir = Files.createTempDir();
+    tmpDir.deleteOnExit();
+
+    File dataDir1 = new File(tmpDir, "data1");
+    File dataDir2 = new File(tmpDir, "data2");
+    String basePath = dataDir1.getAbsolutePath() + "," + dataDir2.getAbsolutePath();
+    shuffleServerConf.setString("rss.storage.type", StorageType.MEMORY_LOCALFILE.name());
+    shuffleServerConf.setString("rss.storage.basePath", basePath);
+    createShuffleServer(shuffleServerConf);
+
+    File dataDir3 = new File(tmpDir, "data3");
+    File dataDir4 = new File(tmpDir, "data4");
+    basePath = dataDir3.getAbsolutePath() + "," + dataDir4.getAbsolutePath();
+    shuffleServerConf.setString("rss.storage.basePath", basePath);
+    shuffleServerConf.setInteger("rss.rpc.server.port", SHUFFLE_SERVER_PORT + 1);
+    shuffleServerConf.setInteger("rss.jetty.http.port", 18081);
+    createShuffleServer(shuffleServerConf);
+
+    File dataDir5 = new File(tmpDir, "data5");
+    File dataDir6 = new File(tmpDir, "data6");
+    basePath = dataDir5.getAbsolutePath() + "," + dataDir6.getAbsolutePath();
+    shuffleServerConf.setString("rss.storage.basePath", basePath);
+    shuffleServerConf.setInteger("rss.rpc.server.port", SHUFFLE_SERVER_PORT + 2);
+    shuffleServerConf.setInteger("rss.jetty.http.port", 17081);
+    createShuffleServer(shuffleServerConf);
+
+    startServers();
+    shuffleServerInfo1 =
+      new ShuffleServerInfo("127.0.0.1-20001", shuffleServers.get(0).getIp(), SHUFFLE_SERVER_PORT);
+    shuffleServerInfo2 =
+      new ShuffleServerInfo("127.0.0.1-20002", shuffleServers.get(1).getIp(), SHUFFLE_SERVER_PORT + 1);
+    shuffleServerInfo3 =
+      new ShuffleServerInfo("127.0.0.1-20003", shuffleServers.get(1).getIp(), SHUFFLE_SERVER_PORT + 2);
+  }
+
+  @Before
+  public void createClient() {
+    // spark.rss.data.replica=3
+    // spark.rss.data.replica.write=2
+    // spark.rss.data.replica.read=2
+    shuffleWriteClientImpl = new ShuffleWriteClientImpl(ClientType.GRPC.name(), 3, 1000, 1,
+      3, 2, 2);
+  }
+
+  @After
+  public void closeClient() {
+    shuffleWriteClientImpl.close();
+  }
+
+  @Test
+  public void readWriteFailTest() throws Exception {
```

Comment:
We should inject more errors, like rpc timeout.

Reply:
ok